### ISSUE TITLE
Responsible Disclosure Policy

### ISSUE THREAD

**Issue by johanndt** [gandrewstone/bitcoin#377]
In light of recent events BU should probably figure out a process for responsible disclosure of exploitable bugs. These should then probably be referenced in CONTRIBUTING.md.

As seen, sensitive fixes can't just be committed directly into the repo as malicious actors are monitoring. However this is an Open Source project and the typical user is highly distrustful of running arbitrary code. I'm not sure how this is handled in other similar projects, but maybe this issue thread could be a good way to explore ideas and suggestions.

My first thoughts are:
1. Have an email listed where anyone can report an exploitable bug.
2. When confirmed, the normal fix -> test -> review -> release process needs to be followed but not in the public repo. Maybe a private repo needs to be forked with access only to the main trusted BU developers.
3. Once a fix is ready, binary builds should be uploaded and a public announcement of a patch release should be made. This announcement should just state that it's a critical exploitable bug and then mention exactly when the fix will be merged into the public repository. It could also state whether or not there are alternate clients, eg Classic or Core which can be run in the interim which might not be affected.
4. At the designated time (24 or 48 hours later), merge the private repo into the public one.

This way people who trust the Unlimited Binaries can upgrade before the exact issue is disclosed. People who'd rather wait to build their own version or see the details can turn off their nodes or switch to something else (temporarily or permanently) and avoid any exploits.

Would love to hear what everyone else thinks.

**Comment by nfujimoto** [comment]
I am also recently thinking about the BU problem. See issues #378. If you agree, please give a vote.

**Comment by ftrader** [comment]
@johanndt Thank you for this thoughtful submission. I would include that the contacted project undertake steps to notify trusted parties at other possibly affected projects, to coordinate a multi-project disclosure and patching processing. As a good example:

https://old.lwn.net/Articles/500444/  (crypt() vulnerability in various BSDs)

> The FreeBSD security team contacted the OpenBSD, NetBSD and DragonflyBSD projects

IMO that article is a must-read / must-think-about one.

For example, of the time taken from discovery to patching in that case was ~3 months, but 7 weeks (!) of those were taken up by efforts to contact other teams. That could probably be improved on if various teams adhered to a common process for notification and further coordination.

I would recommend also extending the process to include filing of a CVE.

**Comment by johanndt** [comment]
@ftrader That's a very good point about vulnerabilities in other libraries. I haven't even considered that.

This raises another issue. What if a problem is found which would affect other upstream or cross-stream repos? Let's say an exploitable issue is found which would affect BU, Classic, Core and XT - since there are large parts of shared code.

I don't really want to pass judgment on the nefariousness of other repo maintainers, but how do we ensure that we do the responsible thing as well without risking "leaks"?

Maybe it is up to the reporting party to inform everyone else?

**Comment by ftrader** [comment]
> Let's say an exploitable issue is found which would affect BU, Classic, Core and XT

This is the situation I was getting at.

Once you've verified the issue on your own codebase, I think it would be a good time to do a cross-check with others' code whether their software is likely affected, and if so, contact them.
I don't see an easy way to solve the trust / leak problem. Maybe others have good ideas on this.
The blockchain could be used to record proof-of-existence for some of the disclosures, but this will not stop leaks as a tactic to hurt other teams. It may take some time to establish the necessary level trust and co-operation between the teams.

A documented common protocol that teams can sign up to would help.

> Maybe it is up to the reporting party to inform everyone else?

Probably to be discussed with them, and handled case by case. I could see projects incentivizing responsible disclosure through bounties, making it lucrative for the researcher to handle this and teams perhaps less willing to share. But there should be a spirit of co-operation between teams which should be re-kindled.

**Comment by roybadami** [comment]
I still have big reservations about releasing binaries before sources.

I don't think most people will want to have to change their workflow in the event of a critical bug, so if our policy is that source will always be delayed 24+ hours in the event of security fixes, that's tantamount to saying that people who need prompt security fixes (which is essentially everyone) should stick to binary releases rather than building from source.

And I'm not sure that's something we want to be telling people.

**Comment by roybadami** [comment]
Further to the above: it would be interesting to get some estimate of how many nodes run binary versus build from source.

In particular, I'd expect that the vast majority of economically important nodes would build from source (and would probably be allergic to running binary releases).  Providing early access to miners/exchanges/payment processors will inevitably have to happen, but it's unlikely that that process can ever catch 100% of businesses, so I think it's perhaps a disservice to such nodes to delay publishing the patches.

**Comment by johanndt** [comment]
@roybadami - after the fallout of yesterday I tend to agree a little. I have to say the fact that uploaded binaries were not GPG signed did not help the situation. My original suggestion assumed that signing the binaries were a given.

Thinking about this some more there are 2 variables you need to consider:

1. Binaries before source: you want to minimize the time between them so people can verify the fixes and binaries are legit.
2. Source before binaries: you want to minimize the time window in which a malicious actor can craft and execute an exploit.

The only logical conclusion is that you need to do them at the exact same time.

Maybe the best course of action then is to revise my original proposal to:

1. Have an email listed where anyone can report an exploitable bug.
2. When confirmed, the normal fix -> test -> review -> release process needs to be followed but not in the public repo. Maybe a private repo needs to be forked with access only to the main trusted BU developers.
3. Once a fix is ready, binary builds should be prepared (but not uploaded yet) and a public announcement of an upcoming patch release should be made. This announcement should just state that it's a critical exploitable bug and then mention exactly when the fix will be available in source and binary form. It could also state whether or not there are alternate clients, eg Classic or Core which can be run in the interim which might not be affected.
4. It might be good to have some official "push" channels for announcements. For example a miner or exchange might want to be emailed about any hot-fixes.
5. At the designated time (a few hours later? enough time for people in different time zones to have reasonably seen the announcement), upload the binaries and merge the private repo into the public one.

In theory it should be quicker to run an upgrade binary than for someone to create an exploit. Eg I can download the binary, install and restart my node in under a minute probably. I can also decide to temporarily disable my node if I want to take my time and test things first.

**Comment by roybadami** [comment]
@johanndt 

> The only logical conclusion is that you need to do them at the exact same time.

That would be my default position.  I do think things are different when the issue is being actively exploited.  In that case, my preference would be to prioritise releasing a source patch ASAP, rather than waiting for signed builds.

**Comment by johanndt** [comment]
> I do think things are different when the issue is being actively exploited.

Agreed. If 60%+ of all BU nodes suddenly go down at once, there is no harm in posting the source patch.

**Comment by ftrader** [comment]
>If 60%+ of all BU nodes suddenly go down at once, there is no harm in posting the source patch.

Agreed, though reality was messy in this recent incident :-) - the source was undergoing patching and review in the private repo, but it already included other issues than the immediate exploit, and notification of other client projects with the full scope of changes had not yet begun when the attack commenced.

It is difficult to decide what the right course of action is if you're attacked at such a moment, because releasing the sources then potentially exposes other clients immediately. A better process would have probably involved those other projects earlier, giving them a clearer chance to patch ahead of release.

Simultaneous public release of binaries and sources for security fixes does sound to me like the best option too.

**Comment by ftrader** [comment]
Related question on bitcoin-dev (concerns CVEs):

https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2017-March/013751.html

**Comment by roybadami** [comment]
Personally I think the right course of action in the most recent indicent would have been to put out a minimal source patch that mitigated the ongoing attack at the earliest possibility opportunity - with binaries following as soon as they could be build by multiple builders.

The full patch and corresponding new binaries could then be put out as a further point release - with simultaneous release as source and binary - as soon as possible after that.
